### PR TITLE
UCT/IB/UD: Drop packets if EP is p2p and disconnected

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Drop packets if EP is p2p and disconnected.

## Why ?

Drop packets on EP which p2p and disconnected to not answer with ACK packets to a peer.
So, a peer will detect an error due to not receiving ACKs.

## How ?

1. Add UD verbs/mlx5 `ep_create` operation to use it for creating EPs and pass initial `flags` for EP (e.g. CONNECT_TO_EP).
2. Updated `uct_ud_ep_process_rx` to drop packets if `UCT_UD_EP_FLAG_DISCONNECTED` and `UCT_UD_EP_FLAG_CONNECT_TO_EP` flags are set on EP.

### CODE REVIEW

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -954,14 +960,14 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
         /* must be connection request packet */
         uct_ud_ep_rx_creq(iface, neth);
         goto out;
-    } else if (ucs_unlikely(!ucs_ptr_array_lookup(&iface->eps, dest_id, ep) ||
-                            (ep->ep_id != dest_id)))
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:958]: this is not needed really, because if we found it in ptr_array it means dest_id is same?
maybe assert it?

File: src/uct/ib/ud/base/ud_ep.c [hunk 2]
```diff
@@ -139,16 +139,14 @@ static void uct_ud_ep_reset(uct_ud_ep_t *ep)
...
+    ucs_time_t diff;
 
     diff = ucs_twheel_get_time(&iface->tx.timer) - ep->close_time;
     if (diff > iface->config.linger_timeout) {
         ucs_debug("ud_ep %p is destroyed after %fs with timeout %fs\n",
                   ep, ucs_time_to_sec(diff),
                   ucs_time_to_sec(iface->config.linger_timeout));
-        ops = ucs_derived_of(iface->super.ops, uct_ud_iface_ops_t);
-        ops->ep_free(&ep->super.super);
+        uct_ud_ep_free(iface, ep);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:149]: it's cleanup - in another PR

File: src/uct/ib/ud/base/ud_ep.c [hunk 3]
```diff
@@ -546,7 +545,8 @@ static ucs_status_t uct_ud_ep_create(uct_ud_iface_t *iface, int path_index,
     uct_ep_h new_ep_h;
     ucs_status_t status;
 
-    status = uct_ep_create(&params, &new_ep_h);
+    status = uct_iface_invoke_ops_func(&iface->super, uct_ud_iface_ops_t,
+                                       ep_create, &params, 0, &new_ep_h);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:549]: this change, and passing flags, seems not needed: we could check params and set CONNECT_TO_EP flag according to params content, after creating the ep

File: src/uct/ib/ud/base/ud_ep.c [hunk 4]
```diff
@@ -397,6 +395,7 @@ UCS_CLASS_INIT_FUNC(uct_ud_ep_t, uct_ud_iface_t *iface,
     ucs_wtimer_init(&self->timer, uct_ud_ep_timer);
     ucs_arbiter_group_init(&self->tx.pending.group);
     ucs_arbiter_elem_init(&self->tx.pending.elem);
+    uct_ud_ep_set_state(self, flags);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:398]: check here:
if (!(params->flags & IFACE_ADDR)) {
 set_state(CONNECT_TO_EP);
}

File: src/uct/ib/ud/base/ud_ep.c [hunk 5]
```diff
@@ -954,14 +960,14 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
-         */
-        ucs_trace("RX: failed to find ep %d, dropping packet", dest_id);
+    } else if (ucs_unlikely(!ucs_ptr_array_lookup(&iface->eps, dest_id, ep))) {
+        ucs_trace("iface=%p: dropping packet dest_id=%u", iface, dest_id);
+        goto out;
+    } else if (ucs_unlikely(ucs_test_all_flags(
+                                    ep->flags,
+                                    UCT_UD_EP_FLAG_DISCONNECTED |
+                                    UCT_UD_EP_FLAG_CONNECT_TO_EP))) {
+        ucs_trace("ep=%p: dropping packet, ep was disconencted", ep);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:970]: `"ep=%p: dropping packet, point-to-point ep was disconencted", ep`

File: src/uct/ib/ud/base/ud_ep.c [hunk 6]
```diff
@@ -954,18 +958,19 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
-    } else if (ucs_unlikely(!ucs_ptr_array_lookup(&iface->eps, dest_id, ep) ||
-                            (ep->ep_id != dest_id)))
-    {
-        /* Drop the packet because it is
-         * allowed to do disconnect without flush/barrier. So it
-         * is possible to get packet for the ep that has been destroyed
-         */
-        ucs_trace("RX: failed to find ep %d, dropping packet", dest_id);
+    } else if (ucs_unlikely(!ucs_ptr_array_lookup(&iface->eps, dest_id, ep))) {
+        ucs_trace("iface=%p: dropping packet dest_id=%u", iface, dest_id);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:962]: minor: "iface %p: dropping packet with dest_id %u"

Comment (dmitrygx) [reply src/uct/ib/ud/base/ud_ep.c]: fixed

**Review by yosefe — APPROVED** [review]

**Review by yosefe — DISMISSED** [review]

File: test/gtest/uct/ib/test_ud.cc [hunk 7]
```diff
@@ -277,6 +278,7 @@ UCS_TEST_SKIP_COND_P(test_ud, flush_ep,
 UCS_TEST_SKIP_COND_P(test_ud, flush_iface,
                      !check_caps(UCT_IFACE_FLAG_AM_SHORT)) {
     connect();
+
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ud.cc:281]: pls revert the unrelated whitespace changes

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Review by yosefe — DISMISSED** [review]

**Review by yosefe — APPROVED** [review]
